src/query/graphite/native/aggregation_functions.go

```diff
@@ -278,6 +281,141 @@ func combineSeriesWithWildcards(
+			end = len(slice)
+		}
+
+		divided = append(divided, slice[i:end])
+	}
+
+	return divided
+}
+
+func evaluateTarget(ctx *common.Context, target string) (ts.SeriesList, error) {
+	eng := NewEngine(ctx.Engine.Storage())
+	expression, err := eng.Compile(target)
+	if err != nil {
+		return ts.NewSeriesList(), err
+	}
+	return expression.Execute(ctx)
+}
+
+/*
+applyByNode takes a seriesList and applies some complicated function (described by a string), replacing templates with unique
+prefixes of keys from the seriesList (the key is all nodes up to the index given as `nodeNum`).
+
+If the `newName` parameter is provided, the name of the resulting series will be given by that parameter, with any
+"%" characters replaced by the unique prefix.
+
+Example:
+
+`applyByNode(servers.*.disk.bytes_free,1,"divideSeries(%.disk.bytes_free,sumSeries(%.disk.bytes_*))")`
+
+Would find all series which match `servers.*.disk.bytes_free`, then trim them down to unique series up to the node
+given by nodeNum, then fill them into the template function provided (replacing % by the prefixes).
+
+Additional Examples:
+
+Given keys of
+
+- `stats.counts.haproxy.web.2XX`
+- `stats.counts.haproxy.web.3XX`
+- `stats.counts.haproxy.web.5XX`
+- `stats.counts.haproxy.microservice.2XX`
+- `stats.counts.haproxy.microservice.3XX`
+- `stats.counts.haproxy.microservice.5XX`
+
+The following will return the rate of 5XX's per service:
+
+`applyByNode(stats.counts.haproxy.*.*XX, 3, "asPercent(%.5XX, sumSeries(%.*XX))", "%.pct_5XX")`
+
+The output series would have keys `stats.counts.haproxy.web.pct_5XX` and `stats.counts.haproxy.microservice.pct_5XX`.
+*/
+func applyByNode(ctx *common.Context, seriesList singlePathSpec, nodeNum int, templateFunction string, newName string) (ts.SeriesList, error) {
+	// using this as a set
+	prefixMap := map[string]struct{}{}
+	for _, series := range seriesList.Values {
+		var (
+			name = series.Name()
+
+			partsSeen int
+			prefix    string
+		)
+
+		for i, c := range name {
+			if c == '.' {
+				partsSeen++
+				if partsSeen == nodeNum+1 {
+					prefix = name[:i]
+					break
+				}
+			}
+		}
+
+		if len(prefix) == 0 {
+			continue
+		}
+
+		prefixMap[prefix] = struct{}{}
+	}
+
+	// transform to slice
+	var prefixes []string
+	for p := range prefixMap {
+		prefixes = append(prefixes, p)
+	}
+	sort.Strings(prefixes)
+
+	var output []*ts.Series
+	maxConcurrency := runtime.NumCPU() / 2
+	for _, prefixChunk := range chunkArrayHelper(prefixes, maxConcurrency) {
+		var (
+			mu       sync.Mutex
+			wg       sync.WaitGroup
+			multiErr xerrors.MultiError
+		)
+
+		for i, prefix := range prefixChunk {
+			_, prefix := i, prefix
+			newTarget := strings.ReplaceAll(templateFunction, "%", prefix)
+			wg.Add(1)
+			go func() {
+				resultSeriesList, err := evaluateTarget(ctx, newTarget)
+
+				if err != nil {
+					mu.Lock()
+					multiErr = multiErr.Add(err)
+					mu.Unlock()
+				}
+
+				for _, resultSeries := range resultSeriesList.Values {
```

Comment:
This will panic if `evaluateTarget` returns an error and `resultSeriesList` is nil

Reply:
ok i am just going to replace 

```
if err != nil {
	mu.Lock()
	multiErr = multiErr.Add(err)
	mu.Unlock()
}
```

with 


```
if err != nil {
	return ts.NewSeriesList(), err
}
```

in that case. And if I'm doing that, then I guess I will get rid of the `multiErr xerrors.MultiError` variable altogether.

Is that OK?

Reply:
Ok, so my interpretation of this comment is that I should change:

```
if err != nil {
	mu.Lock()
	multiErr = multiErr.Add(err)
	mu.Unlock()
}
```

to this
```
if err != nil {
	mu.Lock()
	multiErr = multiErr.Add(err)
	mu.Unlock()
        return
}
```
is that correct?

Reply:
Yeah, exactly :+1: